src/lib/spinel/multi_frame_buffer.hpp

```diff
@@ -309,14 +309,29 @@ template <uint16_t kSize> class MultiFrameBuffer : public FrameWritePointer
      *
      * Saved frame can be retrieved later using `GetNextSavedFrame()`.
      *
+     * @retval OT_ERROR_NONE     Successfully saved the buffer and prepared the write pointer for the next frame.
+     * @retval OT_ERROR_NO_BUFS  Insufficient buffer space.
      */
-    void SaveFrame(void)
+    otError SaveFrame(void)
     {
-        Encoding::LittleEndian::WriteUint16(GetSkipLength() + GetLength(), mWriteFrameStart + kHeaderTotalLengthOffset);
-        mWriteFrameStart = mWritePointer;
-        IgnoreError(SetSkipLength(0));
-        mWritePointer    = GetFrame();
-        mRemainingLength = static_cast<uint16_t>(mBuffer + kSize - mWritePointer);
+        otError error = OT_ERROR_NONE;
+
+        // If the next header will overflow the buffer, we can't save the frame.
+        if (mWritePointer + kHeaderSize > GetArrayEnd(mBuffer))
+        {
+            error = OT_ERROR_NO_BUFS;
+        }
+        else
+        {
+            Encoding::LittleEndian::WriteUint16(GetSkipLength() + GetLength(),
+                                                mWriteFrameStart + kHeaderTotalLengthOffset);
+            mWriteFrameStart = mWritePointer;
+            IgnoreError(SetSkipLength(0));
+            mWritePointer    = GetFrame();
+            mRemainingLength = static_cast<uint16_t>(mBuffer + kSize - mWritePointer);
```

Comment:
```suggestion
```

We can remove the two statements after `SetSkipLength()` now (since they will be set by SetSkipLength()` and we now check that it would not fail).

I think we had these two lines to handle the case where `SetSkipLength(0)` may fail (to make sure frame pointer are correct).